src/renderer/components/kube-object-menu/inject-naive/inject-naive.tsx

```diff
@@ -0,0 +1,54 @@
+/**
+ * Copyright (c) 2021 OpenLens Authors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+import React from "react";
+import { editResourceTab } from "../../dock/edit-resource.store";
+import { hideDetails } from "../../kube-detail-params";
+import { apiManager } from "../../../../common/k8s-api/api-manager";
+import { KubeObjectMenuRegistry } from "../../../../extensions/registries";
+import { getActiveClusterEntity } from "../../../api/catalog-entity-registry";
+import type { KubeObjectMenuDependencies } from "../kube-object-menu";
+import type { KubeObject } from "../../../../common/k8s-api/kube-object";
+
+interface Props {
+  Component: React.ElementType<any>
+  [key: string]: any,
+}
+
+export const InjectNaive = ({ Component, ...props }: Props) => {
+  const kubeObjectMenuRegistry = KubeObjectMenuRegistry.getInstance();
```

Comment:
Not sure if I understand why we need `InjectNaive`, couldn't `KubeObjectMenu` have default props which could be overridden?

Reply:
`InjectNaive` is there to represent the abstraction that should be created between components and it's dependencies. We have seen that usage of default props will limit our choices in the future. Good example would be that dependency is asynchronous.

Of course, in this first step, the `InjectNaive` is just hard-coded solution to get rid of shared global state in UI component itself. In future, we need find a way to inject those dependencies from somewhere else.

Reply:
How could a dependency be asynchronous? React has a synchronous rendering pipeline.

Reply:
Relatively easily. Of course, in React prop should always be synchronous, but with little help of MobX, we can observe the status of a promise and render based on that. You can only imagine the possibilities that it opens. 

However, since it's only one of the benefits, I think it's too deep discussion to have here :) Quick huddle would be great so that we can present it further. 

Our experience says, that creating abstraction here opens a lot of improvements that will make the component itself lot simplier.

Reply:
Indeed, historically asynchronous dependencies have been a wonderful tool to simplify some questionable code here and there. Also the separation of concerns implemented here rubs me the good way.

Let me partake in this huddle, ey guys? :)

Reply:
A huddle might be a good idea. But thinking about this more I still don't see why leveraging default props isn't a better design or at least just as good. That way we don't introduce any new types

Reply:
One thing that comes my mind immediately is that when we get unit testing infrastructure to correct level, we need to start writing unit tests to larger scope to support refactoring in scale. 

That being said, it would cause the issue that you need to provide huge amount of dependencies as props from top level to component levels. 

But yeah, let's schedule huddle and go through it asap.

Reply:
That is why I think we should also leverage default props, so that the users of a component don't need to change

Reply:
How would you handle following situation?

**You want to create unit test for a component which uses other components that have dependencies that need to be stubbed.**

For example:

- Render page in unit test
- One of the component in page has dependency that gets data from an API call
- You want to be able to control the result of said API call and see what effect it has on the page

Disclaimer: 
- API call is just an example and it's stupid one but could be anything that need to be stubbed in test
- jest.doMock works, but performance is horrible and it will cause issues when you have large test suite.
- Shared global state is bad and shouldn't be used here

To solve this, you would need to pass the dependency from unit test to top level (page) component to child component, but if there is an abstraction for injecting said dependency, you can just stub the dependency using the abstraction and all places that inject the dependency will have exactly same instance of it. This makes our unit test suite much more believable and robust.

Reply:
I don't know if I buy the robustness argument since the effect is the same, and is believability really a benefit if you are just hiding where you inject the code?

Reply:
Not hiding of course. As said, it was not ready when we opened the draft for discussion. We are preparing one example how it could be done :)

Reply:
One option to "inject dependencies" is to use custom renderer https://testing-library.com/docs/react-testing-library/setup/#custom-render

There are some real world examples in lenscloud/frontend/__test__. (Maybe not the best one...).

I think it's better to stop using defaultProps, React team seems to plan to retire it for functional components at some point in a major release.

Reply:
Check the proposed solution in #4338.